rslib/src/card_rendering/writer.rs

```diff
@@ -0,0 +1,210 @@
+            Tag::Other(tag) => self.write_other_tag(buf, tag),
+        };
+    }
+
+    fn write_tts_tag(&mut self, buf: &mut String, tag: &TtsTag) {
+        write!(buf, "[anki:tts").unwrap();
+
+        for (key, val) in [
+            ("lang", tag.lang),
+            ("voices", &tag.voices.join(",")),
+            ("speed", &tag.speed.to_string()),
+        ] {
+            self.write_tag_option(buf, key, val);
+        }
+        if let Some(blank) = tag.blank {
+            self.write_tag_option(buf, "cloze_blank", blank);
+        }
+        for (key, val) in &tag.options {
+            self.write_tag_option(buf, key, val);
+        }
+
+        write!(buf, "]{}[/anki:tts]", tag.content).unwrap();
+    }
+
+    fn write_other_tag(&mut self, buf: &mut String, tag: &OtherTag) {
+        write!(buf, "[anki:{}", tag.name).unwrap();
+        for (key, val) in &tag.options {
+            self.write_tag_option(buf, key, val);
+        }
+        buf.push(']');
+        self.write_tag_content(buf, tag.content);
+        write!(buf, "[/anki:{}]", tag.name).unwrap();
+    }
+
+    fn write_tag_option(&mut self, buf: &mut String, key: &str, val: &str) {
+        if val.contains::<&[char]>(&[']', ' ', '\t', '\r', '\n']) {
+            write!(buf, " {}=\"{}\"", key, val).unwrap();
+        } else {
+            write!(buf, " {}={}", key, val).unwrap();
+        }
+    }
+
+    fn write_tag_content(&mut self, buf: &mut String, content: &str) {
+        buf.push_str(content);
+    }
+}
+
+struct AvStripper;
+
+impl AvStripper {
+    fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Write for AvStripper {
+    fn write_sound(&mut self, _buf: &mut String, _ressource: &str) {}
+
+    fn write_tts_tag(&mut self, _buf: &mut String, _tag: &TtsTag) {}
+}
+
+struct AvExtractor<'a> {
+    side: char,
+    tags: Vec<pb::AvTag>,
+    tr: &'a I18n,
+}
+
+impl<'a> AvExtractor<'a> {
+    fn new(question_side: bool, tr: &'a I18n) -> Self {
+        Self {
+            side: if question_side { 'q' } else { 'a' },
+            tags: vec![],
+            tr,
+        }
+    }
+
+    fn write_play_tag(&self, buf: &mut String) {
+        write!(buf, "[anki:play:{}:{}]", self.side, self.tags.len()).unwrap();
+    }
+
+    fn transform_tts_content(&self, tag: &TtsTag) -> String {
+        strip_html_for_tts(tag.content).replace(
+            "[...]",
+            tag.blank.unwrap_or(&self.tr.card_templates_blank()),
+        )
+    }
+}
+
+impl Write for AvExtractor<'_> {
+    fn write_sound(&mut self, buf: &mut String, ressource: &str) {
+        self.write_play_tag(buf);
+        self.tags.push(pb::AvTag {
+            value: Some(pb::av_tag::Value::SoundOrVideo(
+                decode_entities(ressource).into(),
+            )),
+        });
+    }
+
+    fn write_tts_tag(&mut self, buf: &mut String, tag: &TtsTag) {
+        self.write_play_tag(buf);
+        self.tags.push(pb::AvTag {
+            value: Some(pb::av_tag::Value::Tts(pb::TtsTag {
```

Comment:
"blank_cloze" is currently not reported to the frontend. Should it be inserted into `other_args`?

Reply:
Since the replacement is already done at that point, it's probably not too useful. If an add-on wanted to do more elaborate transforms, they'd be better off having the user set the replacement to [...], and then doing the replacement themselves.